### PR TITLE
[XRay][X86] Handle conditional calls when lowering patchable tail calls

### PR DESCRIPTION
**Pull request by rickyz**
xray instruments tail call function exits by inserting a nop sled before
the tail call. When tracing is enabled, the nop sled is replaced with a
call to `__xray_FunctionTailExit()`. This currently does not work for
conditional tail calls, as the instrumentation assumes that the tail
call will be unconditional. This causes two issues:
 - `__xray_FunctionTailExit()` is inappropately called even when the
   tail call is not taken.
 - `__xray_FunctionTailExit()`'s prologue/epilogue adjusts the stack
   pointer with add/sub instructions. This clobbers condition flags,
   which can flip the condition used for the tail call, leading to
   incorrect program behavior.

Fix this by rewriting conditional calls when lowering patchable tail
calls.

With this change, a conditional patchable tail call like:
```
  je target
```

Will be lowered to:
```
  jne .fallthrough
  .p2align 1, ..
.Lxray_sled_N:
  SLED_CODE
  jmp target
.fallthrough:
```

### CODE REVIEW

**Comment by rickyz** [comment]
@llvm/pr-subscribers-xray since the paths at https://github.com/orgs/llvm/teams/pr-subscribers-xray don't seem to match this PR.

**Comment by toddlipcon** [comment]
If the tail call has no instrumentation at all, doesn't the xray-tracked stack end up messed up? ie we think the tail call is nested inside the main call instead of being a sibling, and then potentially never trace an exit for the main call?

**Comment by rickyz** [comment]
Ah, I was hoping that the optionality of https://github.com/llvm/llvm-project/blob/c60aa430dc4085d276a630699323068bf36bd9d8/llvm/lib/CodeGen/XRayInstrumentation.cpp#L39 meant that it was OK to omit trace records, but it does seem like it would lead to inaccurate traces.

Thinking about this more, it should actually be straightforward to rewrite conditional tail calls when lowering them - I've updated this change to do that.

**Comment by rickyz** [comment]
By the way, I'm not sure whether this bug happens on other architectures with xray support. The most straightforward reference I found to emitting conditional tail calls was https://github.com/llvm/llvm-project/blob/c60aa430dc4085d276a630699323068bf36bd9d8/llvm/lib/CodeGen/BranchFolding.cpp#L1536, and `canMakeTailCallConditional()` is only true on x86. I'm not sure if I am missing other places that might generate these though.

File: llvm/test/CodeGen/X86/xray-conditional-tail-call.ll [hunk 1]
```diff
@@ -0,0 +1,35 @@
+; RUN: llc -mtriple=x86_64 < %s | FileCheck %s
```
Comment (MaskRay) [inline llvm/test/CodeGen/X86/xray-conditional-tail-call.ll:1]: Reuse xray-tail-call-sled.ll 

https://maskray.me/blog/2021-08-08-toolchain-testing#i-dont-know-an-existing-test-can-be-enhanced

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> By the way, I'm not sure whether this bug happens on other architectures with xray support. The most straightforward reference I found to emitting conditional tail calls was
> 
> https://github.com/llvm/llvm-project/blob/c60aa430dc4085d276a630699323068bf36bd9d8/llvm/lib/CodeGen/BranchFolding.cpp#L1536
> 
> and `canMakeTailCallConditional()` is only true on x86. I'm not sure if I am missing other places that might generate these though.

Yes. It seems that only x86 performs conditional tail call.

Most RISC architectures cannot perform conditional tail call (except very specific assembly files) because conditional branches usually have a shorter range than unconditional branches and might not reach the destination at link time. RISC architectures typically do not implement range extension thunks for conditional branches.

@deanberris may have an idea why xray handles tail calls specially.

**Comment by rickyz** [comment]
Thanks!

> @deanberris may have an idea why xray handles tail calls specially.
fwiw, I suspect xray needs to instrument tail calls specially so that the correct caller/callee is relationship is maintained in traces - if a tail call were modeled as a function exit followed by a function entry call, then a caller and its tail callee would appear to be siblings rather than parent/child in the call trace info.

**Review by MaskRay — APPROVED** [review]

**Comment by deanberris** [comment]
> Thanks!
> 
> > @deanberris may have an idea why xray handles tail calls specially.
> 
> fwiw, I suspect xray needs to instrument tail calls specially so that the correct caller/callee is relationship is maintained in traces - if a tail call were modeled as a function exit followed by a function entry call, then a caller and its tail callee would appear to be siblings rather than parent/child in the call trace info.

I'm only a few months late to this, but this is precisely why XRay does this. 😄

**Pull request merged**